flink-runtime/src/main/java/org/apache/flink/streaming/api/operators/sorted/state/BatchExecutionInternalTimeServiceManager.java

```diff
@@ -102,14 +113,25 @@ public static <K> InternalTimeServiceManager<K> create(
             ProcessingTimeService processingTimeService,
             Iterable<KeyGroupStatePartitionStreamProvider> rawKeyedStates,
             StreamTaskCancellationContext cancellationContext) {
-        checkState(
-                factory instanceof BatchExecutionKeyedStateBackend,
-                "Batch execution specific time service can work only with BatchExecutionKeyedStateBackend");
+        BatchExecutionKeyedStateBackend<K> theFactory;
+        boolean asyncStateProcessingMode;
+        if (factory instanceof BatchExecutionKeyedStateBackend) {
+            theFactory = (BatchExecutionKeyedStateBackend<K>) factory;
+            asyncStateProcessingMode = false;
+        } else if (factory instanceof AsyncKeyedStateBackendAdaptor) {
+            theFactory =
+                    (BatchExecutionKeyedStateBackend<K>)
+                            ((AsyncKeyedStateBackendAdaptor<K>) factory).getKeyedStateBackend();
```

Comment:
Check `((AsyncKeyedStateBackendAdaptor<K>) factory).getKeyedStateBackend()`  is `BatchExecutionKeyedStateBackend` here?

And update the description of `IllegalStateException` below.

Reply:
Updated accordingly.